src/Components/ServiceScene/Breakdowns/NumericFilterPopoverScene.tsx

```diff
@@ -0,0 +1,508 @@
+import { SceneComponentProps, sceneGraph, SceneObjectBase, SceneObjectState } from '@grafana/scenes';
+import { Button, ClickOutsideWrapper, Field, FieldSet, Input, Label, Select, Stack, useStyles2 } from '@grafana/ui';
+import React from 'react';
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+import { css, cx } from '@emotion/css';
+import { SelectLabelActionScene } from './SelectLabelActionScene';
+import { addNumericFilter, removeFilter, validateVariableNameForField, VariableFilterType } from './AddToFiltersButton';
+import { FilterOp } from '../../../services/filterTypes';
+import { getAdHocFiltersVariable, getValueFromFieldsFilter } from '../../../services/variableGetters';
+import { logger } from '../../../services/logger';
+import { testIds } from '../../../services/testIds';
+
+export interface NumericFilterPopoverSceneState extends SceneObjectState {
+  labelName: string;
+  variableType: VariableFilterType;
+  gt?: number;
+  gte?: boolean;
+  lt?: number;
+  lte?: boolean;
+  fieldType: 'float' | 'duration' | 'bytes';
+  hasExistingFilter?: boolean;
+}
+
+export type NumericFilterPopoverSceneStateTotal =
+  | (NumericFilterPopoverSceneState & FloatTypes)
+  | (NumericFilterPopoverSceneState & DurationTypes)
+  | (NumericFilterPopoverSceneState & ByteTypes);
+
+enum durationUnitValues {
+  ns = 'ns',
+  us = 'µs',
+  ms = 'ms',
+  s = 's',
+  m = 'm',
+  h = 'h',
+}
+
+enum byteUnitValues {
+  B = 'B',
+  KB = 'KB',
+  MB = 'MB',
+  GB = 'GB',
+  TB = 'TB',
+
+  // Supported but potentially confusing values, commented out for now
+  // kB = 'kB',
+  // KiB = 'KiB',
+  // MiB = 'MiB',
+  // GiB = 'GiB',
+  // TiB = 'TiB',
+}
+
+interface FloatUnitTypes {
+  ltu: '';
+  gtu: '';
+}
+
+interface FloatTypes extends FloatUnitTypes {
+  fieldType: 'float';
+}
+
+interface DurationUnitTypes {
+  ltu: durationUnitValues;
+  gtu: durationUnitValues;
+}
+
+interface DurationTypes extends DurationUnitTypes {
+  fieldType: 'duration';
+}
+
+interface ByteUnitTypes {
+  ltu: byteUnitValues;
+  gtu: byteUnitValues;
+}
+
+interface ByteTypes extends ByteUnitTypes {
+  fieldType: 'bytes';
+}
+
+export class NumericFilterPopoverScene extends SceneObjectBase<NumericFilterPopoverSceneStateTotal> {
+  constructor(state: Omit<NumericFilterPopoverSceneStateTotal, 'gtu' | 'ltu'>) {
+    let units: FloatUnitTypes | DurationUnitTypes | ByteUnitTypes;
+    const fieldType: 'float' | 'bytes' | 'duration' = state.fieldType;
+    if (fieldType === 'bytes') {
+      units = { ltu: byteUnitValues.B, gtu: byteUnitValues.B };
+    } else if (fieldType === 'duration') {
+      units = { ltu: durationUnitValues.s, gtu: durationUnitValues.s };
+    } else if (fieldType === 'float') {
+      units = { ltu: '', gtu: '' };
+    } else {
+      throw new Error(`field type incorrectly defined: ${fieldType}`);
+    }
+
+    // @todo - how to avoid type assertion?
```

Comment:
Type guards.

Reply:
Easier said then done, if you have some time to take a look at this we can discuss how to work around this, if we had a hard requirement to not assert here I would likely remove the discriminated union on the fieldType and just make it a string, which might be more likely to introduce bugs in the future then the type assertion.